src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+    Promises are similar to [ref]s – they can store one value. Here is how they
+    differ from [ref]s:
 
-    Lwt also provides the syntax extension {!Pa_lwt} to make code
-    using Lwt more readable.
-*)
+    - A promise might not have a value yet. This is called a {e pending}
+      promise.
+    - Writing a value into a promise is called {e resolving} it.
+    - It's possible to attach {b callbacks} to a pending promise. They will be
+      run when the promise gets resolved.
+    - A promise can be resolved only once. Once a promise has a value, it's
+      immutable.
 
-(** {2 Definitions and basics} *)
+    So, promises are write-once optional references, and when they don't yet
+    have a value, they store a list of callbacks that are waiting for the value.
+
+    Why is this useful?
+
+    The waiting callbacks make promises a nice data type for asynchronous
+    programming. For example, you can ask Lwt to [read] a file. Lwt immediately
+    returns you only a {e promise} for the data.
+
+    You can neglect this promise for a while. You can do some other computation,
+    request more I/O, etc. At some point, you might decide to attach a callback
+    to the [read] promise, maybe several callbacks.
+
+    In the meantime, the [read] operation is running in the background. Once it
+    finishes, Lwt {e resolves} the [read] promise by putting the data into it.
+    Lwt then runs the callbacks you attached.
+
+    One of those might take the data, and ask Lwt to [write] it to STDOUT. Lwt
+    gives you a promise for that, too, and the process repeats.
+
+    Lwt has a small amount of syntactic sugar to make this look as natural as
+    possible:
+
+{[
+let () =
+  Lwt_main.run begin
+    let%lwt data = Lwt_io.(read_line stdin) in
+    let%lwt () = Lwt_io.printl data in
+    Lwt.return ()
+  end
+
+(* ocamlfind opt -linkpkg -package lwt.ppx,lwt.unix echo.ml && ./a.out *)
+]}
+
+    This is all explained in the next sections:
+
+    - {{: #3_Quickstart} Quick start} links these concepts to actual functions
+      in Lwt – the most fundamental ones.
+    - {{: #3_Tutorial} Tutorial} shows how to write examples like the above, and
+      how concurrency happens.
+    - {{: #3_Executionmodel} Execution model} clarifies control flow when using
+      Lwt.
+    - {{: #3_GuidetotherestofLwt} Guide to the rest of Lwt} shows how
+      {e everything} else in Lwt fits into this framework.
+
+    After that is the {{: 2_Fundamentals} reference proper}, which goes into
+    {e painful} levels of detail on every single type and value in this module,
+    [Lwt]. Please be safe, and read only what you need from it :)
+
+    Happy asynchronous programming!
+
+
+
+    {3 Quick start}
+
+    {e All} of Lwt is variations on:
+
+    - {b Promises} of type ['a ]{!Lwt.t} are placeholders for values of type
+      ['a].
+    - {!Lwt.bind} attaches {b callbacks} to promises. When a promise gets a
+      value, its callbacks are called.
```

Comment:
(This goes with a previous comment about callbacks.) It's not clear here that bind attaches callbacks that resolve a promise. These callbacks can have side effects, but they also must (per type discipline) resolve a promise.

Reply:
This was deliberate, because I'm effectively using `Lwt.bind` as a proxy for all those functions that attach callbacks, not being literal about `bind` itself. I don't want it to be clear that `bind`'s callback returns another promise, because not all the other functions have that (`Lwt.map`, `Lwt.on_success`), so it's not actually what I want to be relevant here.

I'm basically looking at `Lwt.bind` as a special case of a more basic mechanism (callbacks), and I believe that mechanism is more important to getting an understanding of Lwt, than the type signature of `bind`.

`bind` is, of course, the most common one. But I think the reader can find its usage and type signature by clicking on the link. There, they will see the promise-valued callback, and see why it is that way.

Further discussion, objections, etc., welcome :)